### ISSUE TITLE
New lint to use Self instead of the type in new methods

### ISSUE THREAD

**Issue by antoyo** [rust-lang-nursery/rust-clippy#1190]
Hello.
It would be nice to have a lint to check if `new` methods have a specific return type annotation of `Self` instead of the referred type.

For instance, with the following code:

``` rust
struct MyStruct {
}

impl MyStruct {
    fn new() -> MyStruct {
        MyStruct {}
    }
}
```

I would like to have a warning about using `MyStruct` instead of `Self`, proposing to write:

``` rust
impl MyStruct {
    fn new() -> Self {
        MyStruct {}
    }
}
```

This lint could be optional to avoid bothering people who prefer to write the real type.

Note that it is not the same as [this lint](https://github.com/Manishearth/rust-clippy/wiki#new_ret_no_self) which checks if the return type is different than `Self`. Here, the types are the same, but the lint would propose to write `Self` instead of the actual type name.

Thanks.

**Comment by Arnavion** [comment]
This was re-filed as #1674 and the lint was implemented as #1965. Perhaps this one should be closed.
